[flang][Driver] Support -print-supported-cpus and associated aliases

tarunprabhu opened the pull request:
The aliases are -mcpu=help and -mtune=help. There is still an issue with the output which prints an example line that references clang. That is not fixed here because it is printed in llvm/MC/SubtargetInfo.cpp. Some more thought is needed to determine how best to handle this.

Fixes #117010

----------------------
Notes for reviewers: The corresponding test in `clang/test/Driver/print-supported-cpus.c` tests two different architectures in the same file. This has been split into two separate tests here to allow the test to be run in cases where only one of the targets has been built.

Commit:
[flang][Driver] Support -print-supported-cpus and associated aliases

The aliases are -mcpu=help and -mtune=help. There is still an issue with the
output which prints an example line that references clang. That is not fixed
here because it is printed in llvm/MC/SubtargetInfo.cpp. Some more thought is
needed to determine how best to handle this.

Fixes #117010

banach-space commented on clang/include/clang/Driver/Options.td:
```diff
@@ -5899,8 +5899,10 @@ def print_enabled_extensions : Flag<["-", "--"], "print-enabled-extensions">,
   MarshallingInfoFlag<FrontendOpts<"PrintEnabledExtensions">>;
-def : Flag<["-"], "mcpu=help">, Alias<print_supported_cpus>;
-def : Flag<["-"], "mtune=help">, Alias<print_supported_cpus>;
+def : Flag<["-"], "mcpu=help">, Alias<print_supported_cpus>,
+  Visibility<[ClangOption, CC1Option, CLOption, FlangOption, FC1Option]>;
+def : Flag<["-"], "mtune=help">, Alias<print_supported_cpus>,
+  Visibility<[ClangOption, CC1Option, CLOption, FlangOption, FC1Option]>;
```
I'd be tempted to extract these to some documented block, similar to https://github.com/llvm/llvm-project/blob/4086ead63c7e0b56b3b07873117bb7ad7a02d41c/clang/include/clang/Driver/Options.td#L3672-L3691.

tarunprabhu replied:
That's a good idea. I have moved the options into such a block. I have begun adding support for `-print-supported-extensions` and `-print-enabled-extensions` as well which can eventually be added to the block.

banach-space commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4417,7 +4417,8 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
       // Use the -mcpu=? flag as the dummy input to cc1.
       Actions.clear();
-      Action *InputAc = C.MakeAction<InputAction>(*A, types::TY_C);
+      Action *InputAc = C.MakeAction<InputAction>(
+          *A, IsFlangMode() ? types::TY_Fortran : types::TY_C);
```
Unrelated change?

tarunprabhu replied:
This is actually needed. 

The `-mcpu=help` argument is used as a "dummy input" (there is a comment a few lines above that says as much). 
The input action created here eventually gets used in `ShouldUseFlangCompiler` which checks if the input type is accepted by `flang`. `flang` does not accept inputs of type `TY_C` (even if they are "dummies"). If `ShouldUseFlangCompiler` returns `false`, a `cc1` command line is created - which is, obviously, not correct.

banach-space commented:
> Notes for reviewers: The corresponding test in clang/test/Driver/print-supported-cpus.c tests two different architectures in the same file. This has been split into two separate tests here to allow the test to be run in cases where only one of the targets has been built.

Wouldn't something like this work: https://github.com/llvm/llvm-project/blob/4086ead63c7e0b56b3b07873117bb7ad7a02d41c/flang/test/Driver/tune-cpu-fir.f90#L1-L3

?

banach-space reviewed (commented):
Thanks, makes sense. From what I can tell, the behaviour is consistent with Clang (i.e. `-print-supported-cpus` invalidates any other action flags).

Just a couple of smallish comments/suggestions

Commit:
Fix case in comment

Commit:
Address reviewer comments.

- Move the options into a commented block. The -print-supported-extensions and
  related options can be moved into this block in a future PR.

- Combine the tests for the different architectures into a single file.

tarunprabhu commented:
> > Notes for reviewers: The corresponding test in clang/test/Driver/print-supported-cpus.c tests two different architectures in the same file. This has been split into two separate tests here to allow the test to be run in cases where only one of the targets has been built.
> 
> Wouldn't something like this work:

Ah! I keep forgetting that that is now possible. Updated the tests to do that.

banach-space commented on flang/test/Driver/print-supported-cpus.f90:
```diff
@@ -0,0 +1,43 @@
+! AARCH64: cortex-a73
+! AARCH64: cortex-a75
+
+! TODO: This is a line that is printed at the end of the output. The full line
+! also includes an example that references clang. That needs to be fixed and a
+! a check added here to make sure that it references flang, not clang.
```
[nit] I would add that bit with clang invocation. Once this is fixed, the test will fail and we will know that we don't need to write a new test (sometimes less is more)

tarunprabhu replied:
That makes sense.  I added that to the test and updated the comment. It will fail when `llvm/MCSubtargetInfo` is fixed.

banach-space reviewed (approved):
LGTM, my final comment is nit - feel free to ignore. Thanks!

Commit:
Update test. This is expected to fail when llvm/MCSubtargetInfo is fixed so it
does not always print clang in the footer of the output.

Commit:
Make the test file valid Fortran

Commit:
Fix typo in comment

jplehr commented:
I believe this results in tests failures on our bot: https://lab.llvm.org/staging/#/builders/105/builds/9796 (and a few others)
